### PR TITLE
ADD: basic 2D map generator for AI

### PR DESCRIPTION
**Pull request by Segelzwerg**

### CODE REVIEW

File: Assets/Scripts/AI/Cartographer.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+            return matrix;
+        }
+
+        /// <summary>
+        /// Creates a map of all static game objects in the scene.
+        /// </summary>
+        private void FillStaticMap()
+        {
+            _gameObjects.AddRange(collection: Object.FindObjectsOfType<MonoBehaviour>().OfType<AiWrapper>().ToList());
```
Comment (iTitus) [inline Assets/Scripts/AI/Cartographer.cs]: You can try to do `FindObjectsOfType<AiWrapper>` as AiWrapper is a `MonoBehaviour` and thus a Unity Object.

File: doc/developer_guide/LevelDesign/index.adoc [hunk 2]
```diff
@@ -0,0 +1,7 @@
+= Level Design
+
+== Required Components for world objects
+
+Each parent object in the world must have a component that implements an `IAWrapper`. This component is needed for the
```
Comment (iTitus) [inline doc/developer_guide/LevelDesign/index.adoc]: Class is called AiWrapper now

File: Assets/Scripts/AI/Cartographer.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        /// <returns>float matrix [x,y, information]</returns>
+        public float[,] MatrixNnReady()
+        {
+            int infoDimension = 0;
+
+            List<MapEntry> mapEntries = _map.ToList();
+            foreach (MapEntry entry in mapEntries.Where(entry => entry.Dimension() > infoDimension))
+            {
+                infoDimension = entry.Dimension();
+            }
```
Comment (iTitus) [inline Assets/Scripts/AI/Cartographer.cs]: isnt there a builtin Max function for LINQ?

File: Assets/Scripts/AI/MapEntry.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        public MapEntry(SortedDictionary<string, float> attributes)
+        {
+            _attributes = attributes;
+        }
+
+        /// <summary>
+        /// The wrap float attributes of an object.
+        /// </summary>
+
+        public List<float> Attributes => _attributes != null ? _attributes.Values.ToList() : new List<float>();
```
Comment (iTitus) [inline Assets/Scripts/AI/MapEntry.cs]: you are recalculating the list every time you query it or its size. 

maybe just use float[] and ditch the SortedDict for memory and performance efficiency

only calculate the float[] once in the AiWrapper and then save it in the MapEntry (which technically might not be necessary you could use a [x,y][dim] here if that works?)

**Review by iTitus — CHANGES_REQUESTED** [review]
If you are just

File: Assets/Scripts/AI/Cartographer.cs (hunk 3 above)
Comment (Segelzwerg) [reply Assets/Scripts/AI/Cartographer.cs]: yes there is.

File: Assets/Scripts/AI/Cartographer.cs (hunk 1 above)
Comment (Segelzwerg) [reply Assets/Scripts/AI/Cartographer.cs]: I will try, but I think I already did it and something went wrong, but can't remember.

File: Assets/Scripts/AI/MapEntry.cs (hunk 4 above)
Comment (Segelzwerg) [reply Assets/Scripts/AI/MapEntry.cs]: As you mentioned in the chat, I think the sorting is necessary so every time the attributes are in the correct dimension. But the dimension calculation performance can be increased.

**Review by iTitus — APPROVED** [review]

**Pull request merged**